### PR TITLE
Implement Dexter entrypoint serialization

### PR DESCRIPTION
**Pull request by sskeirik**
Fixes: #192 
Fixes: #193 

This PR implements:

- abstract entrypoint parameters for the Dexter contract
- a weak abstract entrypoint parameter validation method
- a method to serialize them into concrete Dexter contract parameters

### CODE REVIEW

File: tests/proofs/dexter/dexter.md [hunk 1]
```diff
@@ -76,6 +76,7 @@ Each entrypoint is given a unique abstract parameter type that we use to simplif
 
 ```k
   syntax EntryPointParams
+  syntax Bool ::= validParams(Bool, EntryPointParams) [function]
```
Comment (ehildenb) [inline tests/proofs/dexter/dexter.md:79]:
```suggestion
  syntax Bool ::= validParams(Bool, EntryPointParams) [function, functional]
  // -----------------------------------------------------------------------
  rule validParams(_, _) => false [owise]
```

Comment (ehildenb) [reply tests/proofs/dexter/dexter.md:79]: Another option would be to make it a macro.

File: tests/proofs/dexter/dexter.md [hunk 2]
```diff
@@ -85,13 +86,18 @@ Each entrypoint is given a unique abstract parameter type that we use to simplif
...
+            syntax AddLiquidityParams ::= AddLiquidity(owner              : Address,
+                                                       minLqtMinted       : Int,
+                                                       maxTokensDeposited : Int,
+                                                       deadline           : Timestamp)
+            rule validParams(_IsFA2, AddLiquidity(_Owner,
+                                                   MinLqtMinted,
+                                                   MaxTokensDeposited,
+                                                  _Deadline))
+                 => MinLqtMinted       >=Int 0
+            andBool MaxTokensDeposited >=Int 0
```
Comment (ehildenb) [inline tests/proofs/dexter/dexter.md:100]:
```suggestion
            rule validParams(_IsFA2, AddLiquidity(_Owner,
                                                   MinLqtMinted,
                                                   MaxTokensDeposited,
                                                  _Deadline))
                 => true
              requires MinLqtMinted       >=Int 0
               andBool MaxTokensDeposited >=Int 0
```

Comment (ehildenb) [reply tests/proofs/dexter/dexter.md:100]: This is if you don't have the `macro`, and want to go with functional input ^^^

File: tests/proofs/dexter/dexter.md [hunk 3]
```diff
@@ -271,7 +519,41 @@ We also define a functions that serialize and deserialize our abstract parameter
...
+    => Left Right Right Right Pair To Pair TokensSold Pair MinXtzBought Deadline
+
+  rule #LoadDexterParams(UpdateTokenPool)
+    => Right Left Left Unit
+
+  rule #LoadDexterParams(UpdateTokenPoolInternalFA12(Balance))
+    => Right Left Right Balance
+
+  rule #LoadDexterParams(UpdateTokenPoolInternalFA2(BalanceOfResult))
+    => Right Left Right BalanceOfResult
```
Comment (ehildenb) [inline tests/proofs/dexter/dexter.md:553]: We can either pass a `Bool` to `#LoadDexterParams` or `UpdateTokenPoolInternal`?

File: tests/proofs/dexter/dexter.md [hunk 4]
```diff
@@ -183,10 +238,203 @@ Each entrypoint is given a unique abstract parameter type that we use to simplif
...
+        syntax TokenToTokenParams ::= TokenToToken(outputDexterContract : Address,
+                                                   minTokensBought      : Int,
+                                                   to                   : Address,
+                                                   tokensSold           : Int,
+                                                   deadline             : Timestamp)
+        rule validParams(_IsFA2, TokenToToken(_OutputDexterContract,
+                                               MinTokensBought,
+                                              _To,
+                                               TokensSold,
+                                              _Deadline))
```
Comment (hjorthjort) [inline tests/proofs/dexter/dexter.md:413]: Validate deadline

File: common.md [hunk 5]
```diff
@@ -98,9 +98,10 @@ The internal representation of mutez is bounded.
   // --------------------------------------------
   rule #MutezOverflowLimit => 2 ^Int 63 // Signed 64 bit integers.
 
-  syntax Bool ::= #IsLegalMutezValue(Int) [function]
-  // -----------------------------------------------
-  rule #IsLegalMutezValue(I) => I >=Int 0 andBool I <Int #MutezOverflowLimit
+  syntax Bool ::= #IsLegalMutezValue(Mutez) [function, functional]
+  // -------------------------------------------------
```
Comment (hjorthjort) [inline common.md:102]:
```suggestion
  syntax Bool ::= #IsLegalMutezValue(Mutez) [function, functional]
 // --------------------------------------------------------------
```

File: common.md [hunk 6]
```diff
@@ -275,11 +276,17 @@ The semantics accept timestamps in one of two formats:
...
+
 ```k
   syntax Int ::= #ISO2Epoch(String) [function, hook(TIME.ISO2Epoch)]
   syntax Timestamp ::= #ParseTimestamp(String) [function]
   rule #ParseTimestamp(S) => #Timestamp(#ISO2Epoch(S)) requires findString(S, "Z", 0) >=Int 0
   rule #ParseTimestamp(S) => #Timestamp(String2Int(S)) requires findString(S, "Z", 0) <Int 0
+
+  syntax Bool ::= #IsLegalTimestamp(Timestamp) [function, functional]
+  // ----------------------------------------------------------------
+  rule #IsLegalTimestamp(#Timestamp(I:Int)) => I >=Int 0
```
Comment (hjorthjort) [inline common.md:289]: Should have an `owise` case if we want it to be `functional`, but not sure how to deal with unevaluated functions. Can `#ISO2Epoch` be marked `functional`? Maybe return `-1` for bad values?

Comment (ehildenb) [reply common.md:289]: Should be OK, `#Timestamp` is the only constructor.

**Review by hjorthjort — APPROVED** [review]

**Pull request merged**